src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -0,0 +1,243 @@
+
+        #[cfg(test)]
+        stats.observed_rtt.add(rtt, now);
+
+        if is_back_pressure {
+            inner.had_back_pressure = true;
+        }
+
+        #[cfg(test)]
+        stats.in_flight.add(inner.in_flight, now);
+
+        inner.in_flight -= 1;
+        emit!(AutoConcurrencyInFlight {
+            in_flight: inner.in_flight as u64
+        });
+
+        let rtt = inner.current_rtt.update(rtt);
+        let avg = inner.past_rtt.average();
+
+        if avg == 0.0 {
+            // No past measurements, set up initial values.
+            inner.past_rtt.update(rtt);
+            inner.next_update = now + Duration::from_secs_f64(rtt);
+        } else if avg > 0.0 && now >= inner.next_update {
+            emit!(AutoConcurrencyAveragedRtt {
+                rtt: Duration::from_secs_f64(rtt)
+            });
+
+            #[cfg(test)]
+            {
+                stats.averaged_rtt.add(rtt, now);
+                stats.concurrency_limit.add(inner.current_limit, now);
+            }
+
+            let threshold = avg * THRESHOLD_RATIO;
+
+            // Back pressure responses, either explicit or implicit due
+            // to increasing response times, trigger a decrease in the
+            // concurrency limit.
+            if inner.current_limit > 1 && (inner.had_back_pressure || rtt >= avg + threshold) {
+                // Decrease (multiplicative) the current concurrency limit
+                let to_forget = inner.current_limit / 2;
+                self.semaphore.forget_permits(to_forget);
+                inner.current_limit -= to_forget;
+            }
+            // Normal quick responses trigger an increase in the
+            // concurrency limit. Note that we only check this if we had
+            // requests to go beyond the current limit to prevent
+            // increasing the limit beyond what we have evidence for.
+            else if inner.current_limit < self.max
+                && inner.reached_limit
+                && !inner.had_back_pressure
+                && rtt <= avg
+            {
+                // Increase (additive) the current concurrency limit
+                self.semaphore.add_permits(1);
+                inner.current_limit += 1;
+            }
+            emit!(AutoConcurrencyLimit {
+                concurrency: inner.current_limit as u64,
+            });
+
+            // Reset values for next interval
+            let new_avg = inner.past_rtt.update(rtt);
+            inner.next_update = now + Duration::from_secs_f64(new_avg);
+            inner.current_rtt.reset();
+            inner.had_back_pressure = false;
+            inner.reached_limit = false;
+        }
+    }
+}
+
+impl<L> Controller<L>
+where
+    L: RetryLogic,
+{
+    pub(super) fn adjust_to_response(
+        &self,
+        start: Instant,
+        response: &Result<L::Response, crate::Error>,
+    ) {
+        let is_back_pressure = match response {
+            Ok(_) => false,
```

Comment:
We may want to add a method to `RetryLogic` similar to `should_retry_response` that looks for things like HTTP 429 where we're explicitly being told to slow down. Not necessary for this PR, but could be a good follow-up.

Reply:
Definitely, yes. At one point I had created an `IsBackPressure` trait (or something like that) that would be used in place of `RetryLogic` on response and error types, but discarded that approach.